### ISSUE TITLE
Android should only raise ANR event when the process in ANR state

### ISSUE THREAD

**Issue by LittleXianyu** [getsentry/sentry-android#475]
In class ANRWatchDog, current logic is  that if ActivityManager is not available, we'll still be able to send ANRs. Can we provide configuration scheme to set only raise ANR event when the process in ANR state.
Can we change Logic like this:

`
@Override
  public void run() {
    setName("|ANR-WatchDog|");

    long interval = timeoutIntervalMillis;
    while (!isInterrupted()) {
      boolean needPost = tick.get() == 0;
      tick.addAndGet(interval);
      if (needPost) {
        uiHandler.post(ticker);
      }

      try {
        Thread.sleep(interval);
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        logger.log(SentryLevel.WARNING, "Interrupted: %s", e.getMessage());
        return;
      }

      // If the main thread has not handled ticker, it is blocked. ANR.
      if (tick.get() != 0 && !reported.get()) {
        if (!reportInDebug && (Debug.isDebuggerConnected() || Debug.waitingForDebugger())) {
          logger.log(
              SentryLevel.DEBUG,
              "An ANR was detected but ignored because the debugger is connected.");
          reported.set(true);
          continue;
        }

        // we only raise an ANR event if the process is in ANR state.
        final ActivityManager am =
            (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        boolean isAnr = false;
        if (am != null) {
          final List<ActivityManager.ProcessErrorStateInfo> processesInErrorState =
              am.getProcessesInErrorState();
          // if list is null, there's no process in ANR state.
          if (processesInErrorState == null) {
            continue;
          }
          for (ActivityManager.ProcessErrorStateInfo item : processesInErrorState) {
            if (item.condition == NOT_RESPONDING) {
              isAnr = true;
              break;
            }
          }
        }
        if (!isAnr) {// isAnr must be true
          continue;
        }
        logger.log(SentryLevel.INFO, "Raising ANR");
        final String message =
            "Application Not Responding for at least " + timeoutIntervalMillis + " ms.";

        final ApplicationNotResponding error =
            new ApplicationNotResponding(message, uiHandler.getThread());
        anrListener.onAppNotResponding(error);
        interval = timeoutIntervalMillis;

        reported.set(true);
      }
    }
  }
`

**Comment by marandaneto** [comment]
hey @LittleXianyu thanks for raising that.

what's the use case? are you experiencing issues with the current implementation? can you give some context? thanks.

**Comment by LittleXianyu** [comment]
I just see that there are too many non aggregated ANR reports after my project access sentry. And I find the logic event ActivityManager  is null , sentry will still report ANR.By the way, on different Android versions, the same error will be reported to different issue because the function call stack is a little different . I'm thinking about using 'beforeSend' to do filter, but there are too many. This problem can be done from the client side or the server side to do data aggregation?

**Comment by marandaneto** [comment]
@LittleXianyu got it, the aggregation doesn't work because stack traces are a bit different (on different OS versions), as you said.
using the `beforeSend` would be a way of filtering it.

could you debug it and see why the `ActivityManager` is null? it should not be, maybe there's a specific lifecycle of the application which makes it null and we are not aware of?

Could you share a link of a few ANRs on sentry.io so I can investigate it as well? thanks :)

**Comment by LittleXianyu** [comment]
We built sentry on our server, I'll ask if I can give you an open account. There are maybe hundreds of different ANR reports on our server, how to categorize the known anr problems. one more question, can sentry be configured to sort by the number of events, and is there configuration that only displays reports when the error number above a certain threshold? We just want to focus on the problems with large daily errors , thanks.
![111](https://user-images.githubusercontent.com/8357944/86249708-39814e80-bbe2-11ea-9b7b-cbdfba8f934f.jpg)

**Comment by LittleXianyu** [comment]
I find Sentry can only catch crash on Android UI Thread?

**Comment by marandaneto** [comment]
> There are maybe hundreds of different ANR reports on our server

which version of the SDK are you working with? there were improvements about that.
Please update your issue to use the `bug report` template, or it's hard to identify and reproduce the issues.

https://github.com/getsentry/sentry-android/issues/new/choose

**Comment by marandaneto** [comment]
> I find Sentry can only catch crash on Android UI Thread?

TLDR: No, any thread should work, raise a new issue if you having problems.

**Comment by marandaneto** [comment]
> the sort by Users or by Frequency is not very accurate, and is there configuration that only displays reports when the error number above a certain threshold? We just want to focus on the problems with large daily errors , thanks.

those questions would be better answered either in the [forum](https://forum.sentry.io/) or in the official [sentry repo](https://github.com/getsentry/sentry)

**Comment by marandaneto** [comment]
@LittleXianyu did you get more context about this issue? thanks

**Comment by marandaneto** [comment]
Closing the issue as a part of large repository cleanup, due to it being inactive and/or outdated.
Please do not hesitate to ping me if it is still relevant, and I will happily reopen and work on it.
Cheers!
